File: backend/app/services/fantasy_trade_value.py
"""
Deterministic fantasy trade values (no randomness).

Shared by trade analyzer API routes and ``trade_analyzer_service``.
"""

from __future__ import annotations

import hashlib
from typing import Any, Dict, List, Optional

_POSITION_RANGES: Dict[str, tuple[float, float]] = {
    "QB": (20.0, 45.0),
    "RB": (15.0, 40.0),
    "WR": (12.0, 38.0),
    "TE": (8.0, 25.0),
    "K": (2.0, 6.0),
    "DEF": (3.0, 8.0),
}

_STRONG_OFFENSES = frozenset({"KC", "BUF", "DAL", "SF", "PHI", "MIA", "LAR"})
_WEAK_OFFENSES = frozenset({"WAS", "CHI", "NYG", "CAR"})


def stable_unit(seed: str) -> float:
    """Map *seed* to a stable float in [0, 1)."""
    digest = hashlib.sha256(seed.encode("utf-8")).hexdigest()
    return int(digest[:8], 16) / 0xFFFFFFFF


def _age_multiplier(age: int) -> float:
    if age <= 24:
        return 1.1
    if age <= 27:
        return 1.0
    if age <= 30:
        return 0.95
    return 0.8


def _team_multiplier(team: str) -> float:
    if team in _STRONG_OFFENSES:
        return 1.05
    if team in _WEAK_OFFENSES:
        return 0.95
    return 1.0


def _scoring_multiplier(position: str, scoring_type: str) -> float:
    scoring = (scoring_type or "standard").lower().replace("-", "_")
    if position in {"WR", "TE"}:
        if scoring == "ppr":
            return 1.15
        if scoring in {"half_ppr", "half"}:
            return 1.08
    if position == "RB" and scoring == "standard":
        return 1.1
    return 1.0


def calculate_deterministic_trade_value(
    player: Dict[str, Any],
    *,
    scoring_type: str = "standard",
) -> float:
    """Stable trade value from Sleeper-style player metadata."""
    position = str(player.get("position") or "UNKNOWN").upper()
    age = int(player.get("age") or 27)
    team = str(player.get("team") or "")
    player_key = str(
        player.get("id")
        or player.get("player_id")
        or player.get("name")
        or player.get("full_name")
        or "unknown"
    )

    min_val, max_val = _POSITION_RANGES.get(position, (8.0, 15.0))
    seed = f"{player_key}:{position}:{age}:{team}:{scoring_type}"
    unit = stable_unit(seed)
    base_value = min_val + unit * (max_val - min_val)

    value = (
        base_value
        * _age_multiplier(age)
        * _team_multiplier(team)
        * _scoring_multiplier(position, scoring_type)
    )
    return round(value, 1)


def select_trade_partner(
    teams: List[Dict[str, Any]],
    *,
    seed: str = "",
) -> Optional[Dict[str, Any]]:
    """Pick a stable trade partner from *teams* using *seed*."""
    if not teams:
        return None
    ordered = sorted(teams, key=lambda team: str(team.get("team_id", "")))
    idx = int(stable_unit(seed or "trade_partner") * len(ordered)) % len(ordered)
    return ordered[idx]

File: frontend/src/lib/fantasy-trade-value.ts
import type { LeagueRules } from '@/lib/fantasy-league-rules';

const POSITION_RANGES: Record<string, [number, number]> = {
  QB: [20, 45],
  RB: [15, 40],
  WR: [12, 38],
  TE: [8, 25],
  K: [2, 6],
  DEF: [3, 8],
};

const STRONG_OFFENSES = new Set(['KC', 'BUF', 'DAL', 'SF', 'PHI', 'MIA', 'LAR']);
const WEAK_OFFENSES = new Set(['WAS', 'CHI', 'NYG', 'CAR']);

/** Keep in sync with backend/app/services/fantasy_league_format.py */
export const QB_PREMIUM_SUPERFLEX = 1.4;
export const QB_PREMIUM_2QB = 1.25;
export const TE_SCARCITY_LARGE_LEAGUE = 1.05;
export const TE_SCARCITY_LARGE_WITH_PREMIUM = 1.08;

/** FNV-1a 32-bit — stable in browser without crypto async. */
export function stableUnit(seed: string): number {
  let hash = 2166136261;
  for (let i = 0; i < seed.length; i += 1) {
    hash ^= seed.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return (hash >>> 0) / 0xffffffff;
}

function ageMultiplier(age: number): number {
  if (age <= 24) return 1.1;
  if (age <= 27) return 1.0;
  if (age <= 30) return 0.95;
  return 0.8;
}

function teamMultiplier(team: string): number {
  if (STRONG_OFFENSES.has(team)) return 1.05;
  if (WEAK_OFFENSES.has(team)) return 0.95;
  return 1.0;
}

function scoringMultiplier(position: string, scoringType: string): number {
  const scoring = scoringType.toLowerCase().replace('-', '_');
  if (position === 'WR' || position === 'TE') {
    if (scoring === 'ppr') return 1.15;
    if (scoring === 'half_ppr' || scoring === 'half') return 1.08;
  }
  if (position === 'RB' && scoring === 'standard') return 1.1;
  return 1.0;
}

function tePremiumFromRules(leagueRules?: LeagueRules | null): number {
  if (!leagueRules) return 0;
  const raw = leagueRules.scoring_settings?.raw_settings ?? {};
  for (const key of ['bonus_rec_te', 'rec_te']) {
    const value = Number((raw as Record<string, unknown>)[key]);
    if (Number.isFinite(value) && value > 0) {
      return value;
    }
  }
  const special = leagueRules.scoring_settings?.special_scoring ?? [];
  if (special.some((entry) => /te/i.test(entry) && /premium/i.test(entry))) {
    return 0.5;
  }
  return 0;
}

function formatMultiplier(position: string, leagueRules?: LeagueRules | null): number {
  if (!leagueRules) return 1;

  const pos = position.toUpperCase();
  const teamCount = leagueRules.team_count ?? 12;
  const qbStarters = leagueRules.roster_settings?.positions?.QB ?? 0;
  const hasSuperflex = Boolean(leagueRules.ai_context?.superflex);
  const is2qb = !hasSuperflex && qbStarters >= 2;

  if (pos === 'QB') {
    if (hasSuperflex) return QB_PREMIUM_SUPERFLEX;
    if (is2qb) return QB_PREMIUM_2QB;
    return 1;
  }

  if (pos === 'TE') {
    const tePremium = tePremiumFromRules(leagueRules);
    let mult = 1;
    if (teamCount >= 12) {
      mult = tePremium > 0 ? TE_SCARCITY_LARGE_WITH_PREMIUM : TE_SCARCITY_LARGE_LEAGUE;
    }
    if (tePremium > 0) {
      mult += tePremium;
    }
    return mult;
  }

  return 1;
}

export function calculateDeterministicTradeValue(
  player: {
    id?: string | number;
    player_id?: string | number;
    name?: string;
    position?: string;
    age?: number;
    team?: string;
  },
  leagueRules?: LeagueRules | null
): number {
  const position = (player.position || 'UNKNOWN').toUpperCase();
  const age = player.age ?? 27;
  const team = player.team || '';
  const playerKey = String(
    player.id ?? player.player_id ?? player.name ?? 'unknown'
  );
  const scoringType =
    leagueRules?.scoring_type || leagueRules?.scoring_settings?.type || 'standard';

  const [min, max] = POSITION_RANGES[position] || [8, 15];
  const seed = `${playerKey}:${position}:${age}:${team}:${scoringType}`;
  const unit = stableUnit(seed);
  const baseValue = min + unit * (max - min);

  const value =
    baseValue *
    ageMultiplier(age) *
    teamMultiplier(team) *
    scoringMultiplier(position, scoringType) *
    formatMultiplier(position, leagueRules);

  return Math.round(value * 10) / 10;
}
